roudnas/fit-rozvrh
# adding a service worker to support offline availability + PWA

The caching method used for caching the response of `API_BASE_URL` via the service worker is [_Stale-While-Revalidate_](https://developer.chrome.com/docs/workbox/modules/workbox-strategies/#stale-while-revalidate)

This service worker solution allows the app to work without an internet connection on chromium based browsers, Firefox and iOS Safari _(tested on Linux and iPhone)_.
### To test this yourself:
1. `npm run build`
2. `npm run preview`
3. Open the printed localhost URL in your browser
3.1. check the Network tab: after a second page refresh, you should see two network calls being made to the  `API_BASE_URL` endpoint. The first one, used for actually rendering the UI, was served from the cache and the second one was made _(in a non-blocking manner)_ to the actual API server to update the cache.
4. To really test the offline availability, stop the server and disable any internet access on your machine
5. Reload the localhost URL _(yes, even without the server running)_
6. The app should still load and be usable without any issues

### TODO:
- ~~iOS, for some reason, is not able to load the PWA while being offline, therefore this PR is a draft~~

^^^ **fucking bullshit**, it just started working the day later without any changes made...
I suspect Cloudflare and their cache for the reason it wasn't working yesterday.
I made the locally running app publicly available using [Cloudflare Tunnel](https://developers.cloudflare.com/cloudflare-one/connections/connect-apps/) _(such an op thing btw)_, so I could access it and test it on my iOS phone.
Since it was routed through Cloudflare, their cache probably cached an old _non-functional_ version and was serving that _(I think)_, even though I deleted Safari's web data and cache on my iPhone.
